File: demo/runner/FunctionTestDemo.java
package demo.runner;

import io.substrait.demo.engines.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import java.io.*;
import java.nio.file.*;
import java.util.*;
import org.json.*;

/**
 * Demo runner for function-level compliance tests.
 * 
 * This demonstrates how the new function test framework works alongside
 * the existing TPC-H tests, providing granular function-level testing.
 */
public class FunctionTestDemo {
    
    private static final String TEST_SUITES_DIR = "../test-suites/functions";
    private static final String OUTPUT_DIR = "output";
    private static final ObjectMapper JSON_MAPPER = new ObjectMapper()
        .enable(SerializationFeature.INDENT_OUTPUT);
    
    public static void main(String[] args) {
        System.out.println("╔════════════════════════════════════════════════════════════╗");
        System.out.println("║   Substrait Function-Level Compliance Testing Demo        ║");
        System.out.println("╚════════════════════════════════════════════════════════════╝");
        System.out.println();
        
        try {
            // Create output directory
            Files.createDirectories(Paths.get(OUTPUT_DIR));
            
            // Run tests for each engine
            runEngineTests(new MockDBEngine());
            runEngineTests(new FastDBEngine());
            runEngineTests(new CloudDBEngine());
            runEngineTests(new DuckDBEngine());
            runEngineTests(new PostgreSQLEngine());
            
            // Generate summary report
            generateSummaryReport();
            
            System.out.println("\n✅ Demo completed successfully!");
            System.out.println("📊 Results saved to: " + OUTPUT_DIR + "/");
            System.out.println("🌐 Open dashboard/index.html to view results");
            
        } catch (Exception e) {
            System.err.println("❌ Error running demo: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    private static void runEngineTests(Object engine) throws Exception {
        String engineName = getDisplayEngineName(engine);
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Testing Engine: " + engineName);
        System.out.println("=".repeat(60));
        
        Map<String, TestResults> categoryResults = new HashMap<>();
        
        // Test each function category - scan all available categories
        String[] categories = scanTestCategories();
        
        for (String category : categories) {
            System.out.println("\n📂 Category: " + category);
            TestResults results = runCategoryTests(engine, category);
            categoryResults.put(category, results);
            
            System.out.printf("   Tests: %d | Passed: %d | Failed: %d | Pass Rate: %.1f%%\n",
                results.total, results.passed, results.failed, results.getPassRate());
        }
        
        // Save results
        saveEngineResults(engineName, categoryResults);
    }
    
    private static String[] scanTestCategories() {
        // Scan test-suites/functions directory for all categories
        List<String> categories = new ArrayList<>();
        try {
            Path testSuitesPath = Paths.get(TEST_SUITES_DIR);
            if (Files.exists(testSuitesPath)) {
                try (var stream = Files.list(testSuitesPath)) {
                    stream.filter(Files::isDirectory)
                          .map(p -> p.getFileName().toString())
                          .sorted()
                          .forEach(categories::add);
                }
            }
        } catch (IOException e) {
            System.err.println("⚠️  Error scanning test categories: " + e.getMessage());
            // Fallback to default categories
            return new String[]{"aggregate", "arithmetic", "boolean", "cast", "comparison",
                               "datetime", "string", "window"};
        }
        return categories.toArray(new String[0]);
    }
    
    private static TestResults runCategoryTests(Object engine, String category) {
        // Simulate running tests by parsing .test files
        TestResults results = new TestResults();
        
        try {
            Path categoryDir = Paths.get(TEST_SUITES_DIR, category);
            if (!Files.exists(categoryDir)) {
                System.out.println("   ⚠️  Category directory not found: " + categoryDir);
                return results;
            }
            
            // Count test files
            try (var stream = Files.list(categoryDir)) {
                long testFileCount = stream.filter(p -> p.toString().endsWith(".test")).count();
                
                // Simulate test execution with realistic pass rates
                results.total = (int)(testFileCount * getAvgTestsPerFile(category));
                results.passed = (int)(results.total * getEnginePassRate(engine, category));
                results.failed = results.total - results.passed;
            }
            
        } catch (IOException e) {
            System.err.println("   ❌ Error reading category: " + e.getMessage());
        }
        
        return results;
    }
    
    private static int getAvgTestsPerFile(String category) {
        // Average number of tests per .test file
        switch (category) {
            case "aggregate": return 28;
            case "window": return 32;
            case "cast": return 72;
            case "boolean": return 40;
            case "datetime": return 25;
            case "arithmetic": return 30;
            case "string": return 35;
            case "comparison": return 25;
            case "array": return 25;
            case "struct": return 20;
            case "map": return 20;
            case "json": return 25;
            case "conditional": return 28;
            case "set": return 27;
            case "geospatial": return 20;
            default: return 20;
        }
    }
    
    private static double getEnginePassRate(Object engine, String category) {
        // Simulate different pass rates for different engines and categories
        String engineName = getDisplayEngineName(engine);
        
        switch (engineName) {
            case "MockDB":
                switch (category) {
                    case "aggregate": return 0.75;
                    case "window": return 0.65;
                    case "cast": return 0.80;
                    case "boolean": return 0.90;
                    case "datetime": return 0.70;
                    case "arithmetic": return 0.85;
                    case "string": return 0.88;
                    case "comparison": return 0.92;
                    case "array": return 0.60;
                    case "struct": return 0.55;
                    case "map": return 0.58;
                    case "json": return 0.50;
                    case "conditional": return 0.82;
                    case "set": return 0.65;
                    case "geospatial": return 0.45;
                    default: return 0.70;
                }
            case "FastDB":
                switch (category) {
                    case "aggregate": return 0.92;
                    case "window": return 0.88;
                    case "cast": return 0.95;
                    case "boolean": return 0.98;
                    case "datetime": return 0.85;
                    case "arithmetic": return 0.96;
                    case "string": return 0.94;
                    case "comparison": return 0.98;
                    case "array": return 0.85;
                    case "struct": return 0.80;
                    case "map": return 0.82;
                    case "json": return 0.78;
                    case "conditional": return 0.93;
                    case "set": return 0.88;
                    case "geospatial": return 0.72;
                    default: return 0.90;
                }
            case "CloudDB":
                switch (category) {
                    case "aggregate": return 0.96;
                    case "window": return 0.94;
                    case "cast": return 0.98;
                    case "boolean": return 1.00;
                    case "datetime": return 0.92;
                    case "arithmetic": return 0.99;
                    case "string": return 0.97;
                    case "comparison": return 1.00;
                    case "array": return 0.92;
                    case "struct": return 0.90;
                    case "map": return 0.91;
                    case "json": return 0.88;
                    case "conditional": return 0.97;
                    case "set": return 0.94;
                    case "geospatial": return 0.85;
                    default: return 0.95;
                }
            case "DuckDB":
                switch (category) {
                    case "aggregate": return 0.94;
                    case "window": return 0.91;
                    case "cast": return 0.96;
                    case "boolean": return 0.99;
                    case "datetime": return 0.90;
                    case "arithmetic": return 0.97;
                    case "string": return 0.95;
                    case "comparison": return 0.99;
                    case "array": return 0.90;
                    case "struct": return 0.87;
                    case "map": return 0.88;
                    case "json": return 0.84;
                    case "conditional": return 0.95;
                    case "set": return 0.92;
                    case "geospatial": return 0.78;
                    default: return 0.93;
                }
            case "PostgreSQL":
                switch (category) {
                    case "aggregate": return 0.89;
                    case "window": return 0.84;
                    case "cast": return 0.93;
                    case "boolean": return 0.97;
                    case "datetime": return 0.83;
                    case "arithmetic": return 0.94;
                    case "string": return 0.92;
                    case "comparison": return 0.97;
                    case "array": return 0.81;
                    case "struct": return 0.77;
                    case "map": return 0.75;
                    case "json": return 0.86;
                    case "conditional": return 0.91;
                    case "set": return 0.90;
                    case "geospatial": return 0.68;
                    default: return 0.88;
                }
            default:
                return 0.80;
        }
    }
    
    private static String getDisplayEngineName(Object engine) {
        if (engine instanceof MockDBEngine) return "MockDB";
        if (engine instanceof FastDBEngine) return "FastDB";
        if (engine instanceof CloudDBEngine) return "CloudDB";
        if (engine instanceof DuckDBEngine) return "DuckDB";
        if (engine instanceof PostgreSQLEngine) return "PostgreSQL";
        return engine.getClass().getSimpleName();
    }
    
    private static void saveEngineResults(String engineName, Map<String, TestResults> results) throws IOException {
        JSONObject json = new JSONObject();
        json.put("engine", engineName);
        json.put("timestamp", System.currentTimeMillis());
        
        JSONObject categories = new JSONObject();
        int totalTests = 0;
        int totalPassed = 0;
        
        for (Map.Entry<String, TestResults> entry : results.entrySet()) {
            JSONObject categoryJson = new JSONObject();
            TestResults r = entry.getValue();
            categoryJson.put("total", r.total);
            categoryJson.put("passed", r.passed);
            categoryJson.put("failed", r.failed);
            categoryJson.put("passRate", r.getPassRate());
            categories.put(entry.getKey(), categoryJson);
            
            totalTests += r.total;
            totalPassed += r.passed;
        }
        
        json.put("categories", categories);
        json.put("totalTests", totalTests);
        json.put("totalPassed", totalPassed);
        json.put("overallPassRate", totalTests > 0 ? (totalPassed * 100.0 / totalTests) : 0);
        
        // Save to file
        String filename = OUTPUT_DIR + "/" + engineName + "_function_tests.json";
        Files.writeString(Paths.get(filename), json.toString(2));
    }
    
    private static void generateSummaryReport() throws IOException {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Summary Report");
        System.out.println("=".repeat(60));
        
        JSONArray engines = new JSONArray();
        
        // Read all engine results
        try (var stream = Files.list(Paths.get(OUTPUT_DIR))) {
            stream.filter(p -> p.toString().endsWith("_function_tests.json"))
                  .forEach(path -> {
                      try {
                          String content = Files.readString(path);
                          engines.put(new JSONObject(content));
                      } catch (IOException e) {
                          System.err.println("Error reading " + path + ": " + e.getMessage());
                      }
                  });
        }
        
        // Create summary
        JSONObject summary = new JSONObject();
        summary.put("timestamp", System.currentTimeMillis());
        summary.put("engines", engines);
        summary.put("testSuiteType", "function_tests");
        
        // Save summary
        Files.writeString(Paths.get(OUTPUT_DIR + "/function_tests_summary.json"),
                         summary.toString(2));
        
        Map<String, Object> sharedSummary = buildSharedSummary(engines);
        JSON_MAPPER.writeValue(Paths.get("dashboard/data/summary.json").toFile(), sharedSummary);
        
        // Print summary table
        System.out.println("\n┌─────────────────┬────────┬─────────┬─────────┬───────────┐");
        System.out.println("│ Engine          │ Total  │ Passed  │ Failed  │ Pass Rate │");
        System.out.println("├─────────────────┼────────┼─────────┼─────────┼───────────┤");
        
        for (int i = 0; i < engines.length(); i++) {
            JSONObject engine = engines.getJSONObject(i);
            System.out.printf("│ %-15s │ %6d │ %7d │ %7d │ %8.1f%% │\n",
                engine.getString("engine"),
                engine.getInt("totalTests"),
                engine.getInt("totalPassed"),
                engine.getInt("totalTests") - engine.getInt("totalPassed"),
                engine.getDouble("overallPassRate"));
        }
        
        System.out.println("└─────────────────┴────────┴─────────┴─────────┴───────────┘");
    }
    
    private static Map<String, Object> buildSharedSummary(JSONArray functionEngines) throws IOException {
        Map<String, Object> summary = new LinkedHashMap<>();
        summary.put("lastUpdated", new Date().toInstant().toString());
        summary.put("totalEngines", functionEngines.length());
        
        Map<String, Map<String, Object>> existingSummaryByEngine = new HashMap<>();
        Path existingSummaryPath = Paths.get("dashboard/data/summary.json");
        if (Files.exists(existingSummaryPath)) {
            Map<String, Object> existingSummary = JSON_MAPPER.readValue(existingSummaryPath.toFile(), Map.class);
            List<Map<String, Object>> existingEngines =
                (List<Map<String, Object>>) existingSummary.getOrDefault("engines", Collections.emptyList());
            for (Map<String, Object> engine : existingEngines) {
                existingSummaryByEngine.put(engine.get("engineName").toString(), engine);
            }
        }
        
        List<Map<String, Object>> engines = new ArrayList<>();
        for (int i = 0; i < functionEngines.length(); i++) {
            JSONObject functionEngine = functionEngines.getJSONObject(i);
            String engineName = functionEngine.getString("engine");
            
            Map<String, Object> engineSummary = existingSummaryByEngine.getOrDefault(engineName, new LinkedHashMap<>());
            engineSummary.put("engineName", engineName);
            engineSummary.putIfAbsent("engineVersion", "demo");
            
            Map<String, Object> functions = new LinkedHashMap<>();
            functions.put("passRate", functionEngine.getDouble("overallPassRate"));
            functions.put("totalTests", functionEngine.getInt("totalTests"));
            functions.put("passed", functionEngine.getInt("totalPassed"));
            functions.put("failed", functionEngine.getInt("totalTests") - functionEngine.getInt("totalPassed"));
            engineSummary.put("functions", functions);
            
            if (!engineSummary.containsKey("tpch")) {
                Map<String, Object> emptyTpch = new LinkedHashMap<>();
                emptyTpch.put("passRate", 0.0);
                emptyTpch.put("totalTests", 0);
                emptyTpch.put("passed", 0);
                emptyTpch.put("failed", 0);
                emptyTpch.put("skipped", 0);
                engineSummary.put("tpch", emptyTpch);
            }
            
            engines.add(engineSummary);
        }
        
        summary.put("engines", engines);
        return summary;
    }
    
    static class TestResults {
        int total = 0;
        int passed = 0;
        int failed = 0;
        
        double getPassRate() {
            return total > 0 ? (passed * 100.0 / total) : 0;
        }
    }
}

// Made with Bob

File: demo/runner/function_test_demo.py
#!/usr/bin/env python3
"""
Substrait Function-Level Compliance Testing Demo

This demo showcases the function test framework by:
1. Parsing function test files
2. Generating simulated test results
3. Creating JSON output files
4. Displaying summary statistics
"""

import json
import sys
from pathlib import Path
from datetime import datetime

# Add SDK to path
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "sdk" / "python"))

try:
    from substrait_compliance.function_test_parser import FunctionTestParser
except ImportError:
    print("⚠️  SDK not found. Running in standalone mode...")
    FunctionTestParser = None


class FunctionTestDemo:
    """Demo runner for function tests."""
    
    def __init__(self):
        self.test_suites_dir = Path(__file__).parent.parent.parent / "test-suites" / "functions"
        self.output_dir = Path(__file__).parent.parent / "output"
        self.output_dir.mkdir(exist_ok=True)
        
    def run(self):
        """Run the demo."""
        print("╔════════════════════════════════════════════════════════════╗")
        print("║   Substrait Function-Level Compliance Testing Demo        ║")
        print("╚════════════════════════════════════════════════════════════╝")
        print()
        
        # Test each engine
        engines = ["MockDB", "FastDB", "CloudDB", "DuckDB", "PostgreSQL"]
        all_results = []
        
        for engine_name in engines:
            print(f"\n{'='*60}")
            print(f"Testing Engine: {engine_name}")
            print('='*60)
            
            results = self.test_engine(engine_name)
            all_results.append(results)
            
            # Save engine results
            self.save_engine_results(engine_name, results)
        
        # Generate summary
        self.generate_summary(all_results)
        
        print("\n" + "="*60)
        print("✅ Demo completed successfully!")
        print(f"📊 Results saved to: {self.output_dir}/")
        print("🌐 View results in the output directory")
        print("="*60)
    
    def scan_test_categories(self):
        """Scan test-suites/functions directory for all categories."""
        categories = {}
        
        # Default test counts per category
        default_counts = {
            "aggregate": {"files": 6, "avg_tests": 28},
            "window": {"files": 6, "avg_tests": 32},
            "cast": {"files": 2, "avg_tests": 72},
            "boolean": {"files": 4, "avg_tests": 50},
            "datetime": {"files": 4, "avg_tests": 40},
            "arithmetic": {"files": 11, "avg_tests": 30},
            "string": {"files": 20, "avg_tests": 35},
            "comparison": {"files": 15, "avg_tests": 25},
            "array": {"files": 6, "avg_tests": 25},
            "struct": {"files": 2, "avg_tests": 20},
            "map": {"files": 3, "avg_tests": 20},
            "json": {"files": 2, "avg_tests": 25},
            "conditional": {"files": 2, "avg_tests": 28},
            "set": {"files": 3, "avg_tests": 27},
            "geospatial": {"files": 4, "avg_tests": 20}
        }
        
        if self.test_suites_dir.exists():
            for category_dir in sorted(self.test_suites_dir.iterdir()):
                if category_dir.is_dir():
                    category_name = category_dir.name
                    # Use default if available, otherwise estimate
                    if category_name in default_counts:
                        categories[category_name] = default_counts[category_name]
                    else:
                        # Count actual test files
                        test_files = list(category_dir.glob("*.test"))
                        categories[category_name] = {
                            "files": len(test_files),
                            "avg_tests": 20  # default estimate
                        }
        else:
            # Fallback to defaults if directory doesn't exist
            categories = default_counts
        
        return categories
    
    def test_engine(self, engine_name):
        """Test an engine across all categories."""
        # Scan for all available categories
        categories = self.scan_test_categories()
        
        results = {
            "engine": engine_name,
            "timestamp": int(datetime.now().timestamp() * 1000),
            "categories": {},
            "totalTests": 0,
            "totalPassed": 0
        }
        
        for category, info in categories.items():
            print(f"\n📂 Category: {category}")
            
            # Count actual test files if parser available
            category_dir = self.test_suites_dir / category
            if category_dir.exists():
                test_files = list(category_dir.glob("*.test"))
                num_files = len(test_files)
                
                # Parse files to count tests if parser available
                if FunctionTestParser:
                    parser = FunctionTestParser()
                    total_tests = 0
                    for test_file in test_files:
                        try:
                            suite = parser.parse_file(test_file)
                            total_tests += len(suite.test_cases)
                        except Exception as e:
                            print(f"   ⚠️  Could not parse {test_file.name}: {e}")
                    
                    if total_tests == 0:
                        total_tests = num_files * info["avg_tests"]
                else:
                    total_tests = num_files * info["avg_tests"]
            else:
                total_tests = info["files"] * info["avg_tests"]
            
            # Simulate test execution with realistic pass rates
            pass_rate = self.get_pass_rate(engine_name, category)
            passed = int(total_tests * pass_rate)
            failed = total_tests - passed
            
            results["categories"][category] = {
                "total": total_tests,
                "passed": passed,
                "failed": failed,
                "passRate": round(pass_rate * 100, 1)
            }
            
            results["totalTests"] += total_tests
            results["totalPassed"] += passed
            
            print(f"   Tests: {total_tests} | Passed: {passed} | Failed: {failed} | Pass Rate: {pass_rate*100:.1f}%")
        
        results["overallPassRate"] = round((results["totalPassed"] / results["totalTests"]) * 100, 1) if results["totalTests"] > 0 else 0
        
        return results
    
    def get_pass_rate(self, engine_name, category):
        """Get simulated pass rate for engine/category."""
        rates = {
            "MockDB": {
                "aggregate": 0.75, "window": 0.65, "cast": 0.80, "boolean": 0.90, "datetime": 0.70,
                "arithmetic": 0.85, "string": 0.88, "comparison": 0.92,
                "array": 0.60, "struct": 0.55, "map": 0.58, "json": 0.50,
                "conditional": 0.82, "set": 0.65, "geospatial": 0.45
            },
            "FastDB": {
                "aggregate": 0.92, "window": 0.88, "cast": 0.95, "boolean": 0.98, "datetime": 0.85,
                "arithmetic": 0.96, "string": 0.94, "comparison": 0.98,
                "array": 0.85, "struct": 0.80, "map": 0.82, "json": 0.78,
                "conditional": 0.93, "set": 0.88, "geospatial": 0.72
            },
            "CloudDB": {
                "aggregate": 0.96, "window": 0.94, "cast": 0.98, "boolean": 1.00, "datetime": 0.92,
                "arithmetic": 0.99, "string": 0.97, "comparison": 1.00,
                "array": 0.92, "struct": 0.90, "map": 0.91, "json": 0.88,
                "conditional": 0.97, "set": 0.94, "geospatial": 0.85
            },
            "DuckDB": {
                "aggregate": 0.94, "window": 0.91, "cast": 0.96, "boolean": 0.99, "datetime": 0.89,
                "arithmetic": 0.97, "string": 0.95, "comparison": 0.99,
                "array": 0.88, "struct": 0.84, "map": 0.86, "json": 0.82,
                "conditional": 0.95, "set": 0.91, "geospatial": 0.78
            },
            "PostgreSQL": {
                "aggregate": 0.90, "window": 0.86, "cast": 0.92, "boolean": 0.96, "datetime": 0.85,
                "arithmetic": 0.94, "string": 0.91, "comparison": 0.97,
                "array": 0.82, "struct": 0.78, "map": 0.80, "json": 0.75,
                "conditional": 0.91, "set": 0.87, "geospatial": 0.70
            }
        }
        return rates.get(engine_name, {}).get(category, 0.80)
    
    def save_engine_results(self, engine_name, results):
        """Save engine results to JSON file."""
        filename = self.output_dir / f"{engine_name}_function_tests.json"
        with open(filename, 'w') as f:
            json.dump(results, f, indent=2)
        print(f"\n💾 Saved: {filename}")
    
    def generate_summary(self, all_results):
        """Generate and save summary report."""
        print("\n" + "="*60)
        print("Summary Report")
        print("="*60)
        
        summary = {
            "timestamp": int(datetime.now().timestamp() * 1000),
            "testSuiteType": "function_tests",
            "engines": all_results
        }
        
        # Save summary
        filename = self.output_dir / "function_tests_summary.json"
        with open(filename, 'w') as f:
            json.dump(summary, f, indent=2)
        
        # Print table
        print("\n┌─────────────────┬────────┬─────────┬─────────┬───────────┐")
        print("│ Engine          │ Total  │ Passed  │ Failed  │ Pass Rate │")
        print("├─────────────────┼────────┼─────────┼─────────┼───────────┤")
        
        for result in all_results:
            engine = result["engine"]
            total = result["totalTests"]
            passed = result["totalPassed"]
            failed = total - passed
            pass_rate = result["overallPassRate"]
            
            print(f"│ {engine:15} │ {total:6} │ {passed:7} │ {failed:7} │ {pass_rate:8.1f}% │")
        
        print("└─────────────────┴────────┴─────────┴─────────┴───────────┘")
        
        print(f"\n💾 Saved: {filename}")
        self.generate_shared_summary(all_results)
    
    def generate_shared_summary(self, all_results):
        """Generate shared dashboard summary with function and TPC-H data."""
        demo_root = Path(__file__).parent.parent
        summary_path = demo_root / "dashboard" / "data" / "summary.json"
        leaderboard_path = demo_root / "output" / "leaderboard.json"
        
        existing_summary = {"engines": []}
        if summary_path.exists():
            with open(summary_path, 'r') as f:
                existing_summary = json.load(f)
        
        existing_by_engine = {
            engine["engineName"]: engine
            for engine in existing_summary.get("engines", [])
        }
        
        tpch_by_engine = {}
        if leaderboard_path.exists():
            with open(leaderboard_path, 'r') as f:
                leaderboard = json.load(f)
            for engine in leaderboard.get("engines", []):
                engine_name = engine.get("engineName")
                if engine_name:
                    tpch_by_engine[engine_name] = {
                        "passRate": engine.get("passRate", 0.0),
                        "totalTests": engine.get("totalTests", 0),
                        "passed": engine.get("passed", 0),
                        "failed": engine.get("failed", 0),
                        "skipped": engine.get("skipped", 0)
                    }
        
        merged_engines = []
        for result in all_results:
            engine_name = result["engine"]
            engine_summary = existing_by_engine.get(engine_name, {
                "engineName": engine_name,
                "engineVersion": "demo"
            })
            
            engine_summary["functions"] = {
                "passRate": result["overallPassRate"],
                "totalTests": result["totalTests"],
                "passed": result["totalPassed"],
                "failed": result["totalTests"] - result["totalPassed"]
            }
            
            engine_summary["tpch"] = tpch_by_engine.get(engine_name, engine_summary.get("tpch", {
                "passRate": 0.0,
                "totalTests": 0,
                "passed": 0,
                "failed": 0,
                "skipped": 0
            }))
            merged_engines.append(engine_summary)
        
        shared_summary = {
            "lastUpdated": datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ"),
            "totalEngines": len(merged_engines),
            "engines": merged_engines
        }
        
        with open(summary_path, 'w') as f:
            json.dump(shared_summary, f, indent=2)


def main():
    """Main entry point."""
    demo = FunctionTestDemo()
    demo.run()


if __name__ == "__main__":
    main()

